sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -449,54 +515,137 @@ static RunQueryRequest makeRequest(Query query, @Nullable String namespace) {
 
     @Override
     public void validate(PBegin input) {
-      checkNotNull(getProjectId(), "projectId");
-      checkNotNull(getQuery(), "query");
+      checkNotNull(getProjectId(), "projectId ValueProvider");
+
+      if (getProjectId().isAccessible()) {
+        checkNotNull(getProjectId().get(), "projectId");
```

Comment:
The user has set `projectId`, (to a VP of some sort), so the thing that is wrong now is that they provided an invalid value for the project ID itself. I would phrase the error in that way -- this is an illegal project ID.

So rather than checking that it's not `null`, I would `checkArgument(isValidProject(getProjectId().get()), "Invalid project id: %s", getProjectId().get())` (with some better caching of intermediate variables).

As for how the function `isValidProjectID` is defined, it could be `!= null` or it could be a proper regex based on whatever Google Cloud docs say. I'm not sure what the right thing to do is there.